cpp/src/streaming/cudf/parquet.cpp

```diff
@@ -0,0 +1,215 @@
+    co_await ctx->executor()->schedule();
+    auto size = static_cast<std::size_t>(ctx->comm()->nranks());
+    auto rank = static_cast<std::size_t>(ctx->comm()->rank());
+    auto source = options.get_source();
+    RAPIDSMPF_EXPECTS(
+        source.type() == cudf::io::io_type::FILEPATH, "Only implemented for file sources"
+    );
+    // TODO: To handle this we need the options object to allow int64 num_rows and a
+    // prefix scan across all the ranks of the total number of rows that would be read by
+    // previous ranks.
+    RAPIDSMPF_EXPECTS(
+        size == 1 || !options.get_num_rows().has_value(),
+        "Reading subset of rows not yet supported in multi-rank execution"
+    );
+    // TODO: To handle this we need a prefix scan across all the ranks of the total number
+    // of rows that would be read by previous ranks.
+    RAPIDSMPF_EXPECTS(
+        size == 1 || options.get_skip_rows() == 0,
+        "Skipping rows not yet supported in multi-rank execution"
+    );
+    auto files = source.filepaths();
+    RAPIDSMPF_EXPECTS(files.size() > 0, "Must have at least one file to read");
+    RAPIDSMPF_EXPECTS(
+        files.size() < std::numeric_limits<int>::max(), "Trying to read too many files"
+    );
+    int files_per_rank =
+        static_cast<int>(files.size() / size + (rank < (files.size() % size)));
+    int file_offset = 0;
+    for (auto i = std::size_t{0}; i < rank; i++) {
+        file_offset +=
+            static_cast<int>(files.size() / size + (i < (files.size() % size)));
+    }
+    auto local_files = std::vector(
+        files.begin() + file_offset, files.begin() + file_offset + files_per_rank
+    );
+    int files_per_split = 1;
+    // TODO: Handle case where multiple ranks are reading from a single file.
+    // TODO: We could be smarter here, suppose that the number of files we end up wanting
+    // is one, but each file is marginally larger than our target_rows_per_chunk, we'd end
+    // up producing many small chunks.
+    if (files_per_rank > 1) {
+        // Figure out a guesstimated splitting.
+        auto source = cudf::io::source_info(local_files[0]);
+        auto metadata = cudf::io::read_parquet_metadata(source);
+        auto const rg = metadata.rowgroup_metadata();
+        auto const num_rows = metadata.num_rows();
+        files_per_split =
+            std::max(static_cast<int>(num_rows_per_chunk / num_rows), files_per_split);
+    }
+    std::vector<Node> read_tasks;
+    auto options_skip_rows = options.get_skip_rows();
+    auto options_num_rows =
+        options.get_num_rows().value_or(std::numeric_limits<int64_t>::max());
+    std::uint64_t sequence_number = 0;
+    if (options_num_rows == 0) {
+        if (rank == 0) {
+            // Rank zero sends empty table of the correct shape/schema. Everyone else
+            // sends nothing.
+            cudf::io::parquet_reader_options empty_opts{options};
+            cudf::io::source_info source{files[0]};
```

Comment:
nit. Unused?

Reply:
nit. And isnt `empty_opts` redundant here?

Reply:
Ah yes, I should write `empty_opts.set_source(source)`